### ISSUE TITLE
Orientation of DICOM pixel arrays

### ISSUE THREAD

**Issue by migonch** [neuro-ml/dicom-csv#8]
get_orientation_matrix сейчас возвращает матрицу A размера 3x3, строки которой - это коэффициенты базисных векторов, связанных со срезом (e_x' - вектор вдоль первой строки среза, e_y' - вектор вдоль первого столбца среза, e_z' - их векторное произведение) в базисе, связанном с пациентом (e_x торчит из левого виска, e_y торчит из спины, e_z торчит из макушки). Эта матрица всегда ортогональная. Матрица A - это матрица перехода из базиса e' в e (e = e'A), а ее транспонированная - из e в e' [вики](https://ru.wikipedia.org/wiki/%D0%9C%D0%B0%D1%82%D1%80%D0%B8%D1%86%D0%B0_%D0%BF%D0%B5%D1%80%D0%B5%D1%85%D0%BE%D0%B4%D0%B0).

Насколько я понимаю, наша цель научится собирать картинку из аксиальных срезов, которые шли бы от головы к ногам, а ориентация пациента на одном срезе у нас пока не стандартизована, но я считаю, что на самом деле она должна быть стандартизована, хотя бы с точностью до поворота (НЕ должно быть чтобы сердце на снимках было слева, а печень справа, потому что насколько я понимаю врачи привыкли что сердце на срезах справа, а печень слева).

Сейчас в dicom-csv картинка собирается не так. Срезы сейчас упорядочиваются по slice location, которое вычисляется, как скалярное произведение поля ImageOrientationPatient (которое содержит координаты (x, y, z) левого верхнего вокселя среза в базисе (e_x, e_y, e_z)) на третью строчку матрицы A. Это дает координату z' этого вокселя вдоль вектора e_z'. Направление этого вектора относительно пациенту вообще говоря произвольное, и в `/nmnt/x3-hdd/production/pacs` есть примеры, где этот вектор направлен как от головы к ногам, так и наоборот. Поэтому упорядочивать срезы вдоль него нет смысла.

Еще один баг, который сейчас есть в dicom-csv: функция get_flipped_axes сейчас возвращает [0], если матрица A = 
-1 0 0
  0 1 0
  0 0 1
хотя, в этом случае индекс вдоль первой строки увеличивается от левого виска к правому (cердце на срезе будет слева), а индекс вдоль первого столбца увеличивается от груди к спине, и в таком случае картинку надо флипать вдоль axis=1.

Что я предлагаю:

1. Если третья строчка матрицы A не равна (0, 0, 1) или (0, 0, -1), это значит, что срезы идут не ортогонально вектору e_z, то есть они не аксиальные. Такое вроде бывает очень редко и такие картинки надо игнорить.
2. Если срезы аксиальные, то в качестве slice location надо просто брать ImagePositionPatient[2] = координата z вдоль вектора e_z, который всегда направлен от ног к голове. Упорядочивать срезы надо чтобы эта координата убывала.
3. В плоскости Oxy такой алгоритм: 
а) если матрицa A[:2, :2] содержит два нулевых элемента вне диагонали, то срезы надо транспонировать
b) затем, если на диагонали (1, 1), то все хорошо, если (-1, 1), то надо флипнуть вдоль axis=-1, если (1, -1), то хз (или флипнуть вдоль axis=0, то есть перевернуть с живота на спину, или флипнуть вдоль axis=1, то есть оставить лежать на животе) и если (-1, -1), то оставить лежать на животе.

**Comment by kurmukovai** [comment]
> Насколько я понимаю, наша цель научится собирать картинку из аксиальных срезов

Нет, в случае c КТ действительно, почти всегда речь идет об _аксиальных_ срезах. В случае с МРТ это не так, есть много _саггитальных_ картинок, и какое-то число _коронарных_. Что происходит с УЗИ я вообще пока не понимаю, но там вроде просто серия изображений без привязки к какой бы то ни было проекции (так что не понятно насколько там целесообразны поля вроде slice location, image orientation и некоторые другие).

**Comment by migonch** [comment]
пункт 1: после сегодняшнего обсуждения с Максом, я немного переписал spatial, и там сейчас собираются серии с любыми срезами (не только с аксиальными)
пункт 2: вычисление slice_locations я оставил как было. Но в order_series нельзя сортировать по старым slice_locations, поэтому сейчас я сортирую по image_position_patient[i], где i - индекс базисного вектора, связанного с пациентом, который лучше всего сонаправлен с нормалью к срезам.
пункт 3: см. решение которое сейчас в spatial.py
